modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/classloader/AbstractArtifactClassLoaderConfigurationAssembler.java

```diff
@@ -0,0 +1,166 @@
+import org.slf4j.Logger;
+
+import java.io.File;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Assembles the class loader configuration for an artifact given all its pieces.
+ */
+public abstract class AbstractArtifactClassLoaderConfigurationAssembler {
+
+  private static final Logger LOGGER = getLogger(AbstractArtifactClassLoaderConfigurationAssembler.class);
+
+  protected org.mule.tools.api.classloader.model.ClassLoaderModel packagerClassLoaderModel;
+
+  public AbstractArtifactClassLoaderConfigurationAssembler(org.mule.tools.api.classloader.model.ClassLoaderModel packagerClassLoaderModel) {
+    this.packagerClassLoaderModel = packagerClassLoaderModel;
+  }
+
+  public ClassLoaderModel createClassLoaderModel() {
+    ArtifactClassLoaderConfigurationBuilder artifactClassLoaderConfigurationBuilder =
+        new ArtifactClassLoaderConfigurationBuilder(packagerClassLoaderModel, getProjectFolder());
+
+    artifactClassLoaderConfigurationBuilder
+        .exportingPackages(getExportedPackages())
+        .exportingResources(getExportedResources());
+    // TODO: add privileged packages and resources
+
+    // TODO: analyze patched dependencies in classloader-model-patch.json
+
+    final List<URL> dependenciesArtifactsUrls =
+        loadUrls(getProjectFolder(), getBundleDependencies(), artifactClassLoaderConfigurationBuilder);
+    dependenciesArtifactsUrls.forEach(artifactClassLoaderConfigurationBuilder::containing);
+
+    // TODO: check if it belongs to the deny-list to decide whether local packages should be populated
+    populateLocalPackages(packagerClassLoaderModel, artifactClassLoaderConfigurationBuilder);
+
+    artifactClassLoaderConfigurationBuilder.dependingOn(new HashSet<>(getBundleDependencies()));
+
+    return artifactClassLoaderConfigurationBuilder.build();
+  }
+
+  private List<URL> loadUrls(File artifactFile,
+                             List<BundleDependency> dependencies,
+                             ArtifactClassLoaderConfigurationBuilder artifactClassLoaderConfigurationBuilder) {
+    final List<URL> dependenciesArtifactsUrls = new ArrayList<>();
+
+    // TODO: consider artifact patches for the case this is run within a Runtime
+
+    final URL artifactFileUrl = getUrl(artifactFile, artifactFile);
+    dependenciesArtifactsUrls.add(artifactFileUrl);
+
+    dependenciesArtifactsUrls.addAll(addArtifactSpecificClassLoaderConfiguration(artifactClassLoaderConfigurationBuilder));
+    dependenciesArtifactsUrls.addAll(addDependenciesToClasspathUrls(artifactFile, dependencies));
+
+    return dependenciesArtifactsUrls;
+  }
+
+  protected Collection<URL> addArtifactSpecificClassLoaderConfiguration(ArtifactClassLoaderConfigurationBuilder artifactClassLoaderConfigurationBuilder) {
+    return Collections.emptyList();
```

Comment:
static import

Reply:
the signature of this method is confusing... does it modify the builder it recieves? Use it to calculate the data to return? both? from the impl in the subclass it seems to be doing both, meke this clear in the javadoc

Reply:
Done.